### DISCUSSION TITLE
MVU + C# Markup: how do I expose internal model property as a public control property?

### DISCUSSION THREAD

**Discussion by kucint** [unoplatform/uno#16278] [Q&A]
Hi,

Basically I want to mimic [ControlLibrary sample code](https://github.com/unoplatform/Uno.Samples/blob/master/UI/ControlLibrary) project using MVU + C# Markup.

I do the following
First I do create my "user control" and the model:

```
public sealed partial class MyUserControl : UserControl
{
    public MyUserControl() => this
        .DataContext(new BindableMyModel(), (ctrl, vm) => ctrl
        .Content( ... ));
}

internal partial record MyModel
{
    public IState<string> Label => State<string>.Value(this, () => "This is my Label");

    //... other states follow here
};
```

Now I want to expose the Label Property, that it can be modified outside of the control.
Following [ControlLibrary sample code](https://github.com/unoplatform/Uno.Samples/blob/master/UI/ControlLibrary) I create a Label Property in  `MyUserControl` class:

```
    public string Label
    {
        get { return (string)GetValue(LabelProperty); }
        set { SetValue(LabelProperty, value); }
    }

    public static readonly DependencyProperty LabelProperty =
        DependencyProperty.Register(
            nameof(Label),
            typeof(string),
            typeof(MyUserControl),
            new PropertyMetadata(
                "label default",
                (s, e) => ((MyUserControl)s)?.OnLabelChanged(e))
        );

    private void OnLabelChanged(DependencyPropertyChangedEventArgs e)
    {
        System.Diagnostics.Debug.WriteLine($"Label Changed from [{e.OldValue}] to [{e.NewValue}]");
    }
```

**Question**: How am I supposed to bind this public property to my internal model?

**Comment by dr1rrb** [comment]
Well the sample if about creating a `Control` but in your code snippets you are creating a `UserControl`.

It's uncommon to have `DependencyProperty` on a `UserControl`. Could you give us more details (a repro app?) on what you are trying to do ?

**Comment by kucint** [comment]
OK. that is true that my snippet is a `UserControl` and the the sample is a `Control`. I was thinking, it plays no role.
if `DependencyProperty` is not way to go, what is?

So let reformulate and split the question into two topics:
**Q1**: How do I expose internal model property to the consumer in templated `Control`?
**Q2**: How do I expose internal model property to the consumer in `UserControl`?

**Comment by kucint** [comment]
@dr1rrb 
In general I intend to create couple of custom controls (both user and templated) that:
- shall be configurable, it means consumer can change the way the controls behave
- shall allow the consumer to access the state of a control (public states) that is related to a combination of (many) internal model properties (internal states).
What is the best way to do this for `UserControl` and for `Control` ?
